File: Closures/Attribute_List.java
import java.util.ArrayList;



// 属性列表，将所有属性保存在其中
public class Attribute_List {
    
    // 使用数组存储
    public ArrayList<Attribute>attributes;

    // 构造函数,初始化属性数组
    public Attribute_List(){
        attributes=new ArrayList<Attribute>();
    }

    // 判断两个属性数组是否相同：
    public boolean is_equal(Attribute_List other)
    {
        // 判断属性组长度是否相同
        int length0=this.attributes.size();
        int length1=other.attributes.size();
        // 数组长度不相同直接返回false
        if (length0!=length1)return false;

        for(int i=0;i<length0;i++){
            Attribute a=this.attributes.get(i);
            Attribute b=other.attributes.get(i);
            // 假如字符串不相等，直接返回false
            if (!a.get_name().equals(b.get_name()))
            return false;
        }
    // 全部相同直接返回真
       return true;
    }

    // 获取属性列表的index位置的属性
    public  Attribute get_item(int index){
        return this.attributes.get(index);
    }


    // 将属性列表内容输出
    public String toString() 
    {
        String str="{";
        for(int i=0;i<this.attributes.size();i++){
             str+=this.attributes.get(i).get_name();
             str+=",";
        }
        str=str.substring(0,str.length()-1);
        str+="}";

        // 返回值,example{A,B,C,D,E}
        return str;
    }

    // 删除元素
    public boolean remove(Attribute r){
        
        // 删除该元素
        return attributes.remove(r);
    }


    // 查找下标
    public int index(Attribute r){

        // 返回位置
        return attributes.indexOf(r);
    }


    // 加入一个新属性
    public void append(Attribute r){
        // 添加
        this.attributes.add(r);
    }

    //加入一个属性列
    public void append_list(Attribute_List r) {
        for (int i=0;i<r.attributes.size();i++){
            // 将一个属性列表加入到现有的属性列表中
            this.attributes.add(r.attributes.get(i));
        }
    }

    // 获得所有的属性列
    public ArrayList<Attribute> get_attributes(){
        return this.attributes;
    }

    // 属性值保存
    public void copy(Attribute_List r){
        this.attributes=new ArrayList<Attribute>(r.attributes);
    }

    // 属性值复制
    public Attribute_List clone(){
        Attribute_List new_a=new Attribute_List();
        new_a.attributes=new ArrayList<Attribute>(this.attributes);
        return new_a;
    }

    // 根据属性值String进行排序
    public void sort(){
        this.attributes.sort(new SortByIndex());
    }

    // 获取属性集的长度
    public int count(){
        return this.attributes.size();
    }

}

File: example/attribute_list.py
# 属性列表类，是属性组AttributeGroup和AttributeSet的基类
class AttributeList:
    def __init__(self):
        self.attributes = []

    def __getitem__(self, item):
        return self.attributes[item]

    def __eq__(self, other):
        c1 = len(self.attributes)
        c2 = len(other.attributes)
        if c1 != c2:
            return False
        else:
            for i in range(0, c1):
                if self.attributes[i] != other.attributes[i]:
                    return False
                else:
                    continue
            return True

    def __ne__(self, other):
        c1 = len(self.attributes)
        c2 = len(other.attributes)
        if c1 != c2:
            return True
        else:
            for i in range(0, c1):
                if self.attributes[i] != other.attributes[i]:
                    return True
                else:
                    continue
            return False

    def __str__(self):
        s = '{'
        for c in self.attributes:
            s += c.get_name()
            s += ','
        s = s.rstrip(',') + '}'
        return s

    def remove(self, attribute):
        self.attributes.remove(attribute)

    def index(self, attribute):
        return self.attributes.index(attribute)

    def append(self, attribute):
        self.attributes.append(attribute)

    def append_all(self, attribute_set):
        for c in attribute_set.attributes:
            self.attributes.append(c)

    def get_attributes(self):
        return self.attributes

    def copy(self, other):
        self.attributes = other.attributes.copy()

    def clone(self):
        a = AttributeList()
        a.attributes = self.attributes.copy()
        return a

    def sort(self):
        self.attributes.sort()

    def count(self):
        return len(self.attributes)
